### ISSUE TITLE
Vector tile does not render when zoom level greater than 14

### ISSUE THREAD

**Issue by giswqs** [python-visualization/folium#1715]
**Describe the bug**
The vector tile layer does not render when zoom level is great than 14.
@iwpnd

**To Reproduce**
https://nbviewer.org/github/python-visualization/folium/blob/main/examples/plugin-vector-tiles.ipynb

```python
import folium
from folium.plugins import VectorGridProtobuf
url = "https://vectortiles3.geo.admin.ch/tiles/ch.swisstopo.leichte-basiskarte.vt/v1.0.0/{z}/{x}/{y}.pbf"
m = folium.Map(location=[46.8, 8.2], zoom_start=14)

options = {'maxNativeZoom': 20, 'maxZoom': 20}
# options = {'max_native_zoom': 20, 'max_zoom': 20} #not working either

vc = VectorGridProtobuf(url, "vector tile", options)
m.add_child(vc)
m
```
![Peek 2023-01-27 09-38](https://user-images.githubusercontent.com/5016453/215112856-da2acaf2-3b0f-472e-a310-3489db7f472c.gif)

**Expected behavior**
The vector tile layer should render in a zoom level much larger than 14 so that we can see the details of vector data. 

**Environment (please complete the following information):**
 - Python version: 3.9
 - folium version: 0.13.0

**Possible solutions**
The leaflet `VectorGrid.Protobuf` example appears to have the `maxNativeZoom` parameter. I am not sure if the folium plugin can pass the `maxNativeZoom` and `maxZoom` parameters to  `VectorGrid.Protobuf`.

https://leaflet.github.io/Leaflet.VectorGrid/vectorgrid-api-docs.html
```javascript
L.vectorGrid.protobuf("https://free-{s}.tilehosting.com/data/v3/{z}/{x}/{y}.pbf.pict?key={key}", {
    vectorTileLayerStyles: { ... },
    subdomains: "0123",
    key: 'abcdefghi01234567890',
    maxNativeZoom: 14
}).addTo(map);
```

**Relevant issues**
https://github.com/microsoft/PlanetaryComputer/discussions/175 @TomAugspurger

**Comment by TomAugspurger** [comment]
One quick note: I think we only provide vector tiles for certain zoom levels. Maybe @mmcfarland knows the maximum zoom level off the top of his head.

I believe in our explorer we render the tiles for the largest zoom level, even when the user zooms in past that.

**Comment by giswqs** [comment]
@TomAugspurger Yeah, the issue occurs with all vector tile layers, not just the Planetary Computer vector tiles. So I guess the issue is with the plugin itself rather than the data.

**Comment by Conengmo** [comment]
Could you verify that folium adds the zoom option correctly in the html? If you save your map as html, check that it looks like in the Leaflet example. If that looks correct, then we know the issue is upstream. Otherwise, there's something wrong on how folium processes the options.

**Comment by giswqs** [comment]
It appears that the zoom options can be found in the html, so this might be an upstream issue. 

![image](https://user-images.githubusercontent.com/5016453/215132208-9b2f3cc1-bea9-419a-8540-b295001f8085.png)

**Comment by giswqs** [comment]
I found the `maxNativeZoom` [issue](https://github.com/Leaflet/Leaflet.VectorGrid/issues/74) in the upstream [Leaflet.VectorGrid](https://github.com/Leaflet/Leaflet.VectorGrid/issues/74). Unfortunately, the repo is no longer being actively maintained and has no update since Sept 2021. It is unlikely that the issue will be resolved any time soon. 

There is an alternative called [`Leaflet.VectorTileLayer`](https://gitlab.com/jkuebart/Leaflet.VectorTileLayer), which is being actively developed by @jkuebart. It appears to support `min/maxNativeZoom`.

It would be great if folium can switch to use `Leaflet.VectorTileLayer`.

The beauty of vector tile is that you can zoom way in to see the details of the shapes. Without the `maxNativeZoom` and `maxZoom`, the vector tile rendering is not very useful.

**Comment by iwpnd** [comment]
Nothing I can do here as this is not a bug or an issue at all but a deliberate decision to not render more than the maximum zoom levels the provider offers.
I reckon using `maxNativeZoom` in the doc is a mistake rather than pointing to the feature you're expecting. 

I chose VectorGrid because it was the supported plug-in for vector tiles under the umbrella of leaflet. The one you're providing now is, while actively maintained, a one man show.

> The beauty of vector tile is that you can zoom way in to see the details of the shapes.

Depending on the configuration, zoom level and provider, features in a tile are simplified on creation, or even excluded entirely. As the client does not have access to this information it's best to stick to min/max provided. Everything after that is an approximation.

That said, I can see the appeal for the client to decide it. I'm discovering the plug-in you mentioned and am confident to provide it to folium as well. 
I would provide you with an external repository you can add to folium and test, and if it covers your use case would create a PR with folium. Whether to keep both or drop one is up to folium after that.

**Comment by giswqs** [comment]
@iwpnd Thank you for your help. I am specifically interested in visualizing the [Microsoft Building Footprint data](https://planetarycomputer.microsoft.com/explore?c=-86.2241%2C39.7737&z=13.31&v=2&d=ms-buildings&m=Global&r=Default&s=true%3A%3A100%3A%3Atrue&sr=desc&ae=0) on Planetary Computer. I think the provider does provide a maximum zoom level larger than 14, as you can zoom way in to see the details of the building footprints. 

![image](https://user-images.githubusercontent.com/5016453/215270935-95c49d2e-0f99-4fac-8f58-3e9eb11a395a.png)

And this is what it renders in folium. If you further zoom in, the tile does not render and disappears. 
```python
import folium
from folium.plugins import VectorGridProtobuf
url = "https://planetarycomputer.microsoft.com/api/data/v1/vector/collections/ms-buildings/tilesets/global-footprints/tiles/{z}/{x}/{y}"
m = folium.Map(location=[40, -100], zoom_start=13)
options = {'maxNativeZoom': 20, 'maxZoom': 20}
vc = VectorGridProtobuf(url, "vector tile", options)
m.add_child(vc)
m
```

![image](https://user-images.githubusercontent.com/5016453/215270968-61c67928-76b9-4403-98e0-4883d8a83d86.png)

**Comment by iwpnd** [comment]
There is a difference as to the maximum zoom the provider offers, and the maximum zoom the client "stretches" the result of the last maximum zoom to the requested zoom.

Take for instance Milano:

While [12/2152/5862](https://planetarycomputer.microsoft.com/api/data/v1/vector/collections/ms-buildings/tilesets/global-footprints/tiles/12/2152/5862) returns a tile, [14/8610/5862](https://planetarycomputer.microsoft.com/api/data/v1/vector/collections/ms-buildings/tilesets/global-footprints/tiles/14/8610/5862) does not. But the viewer you provide takes care, that the features of the last real tile, are set to the clients requested zoom level.

L.VectorGrid does not have that capability, L.VectorTileLayer on the other hand seems to have that feature.

**Comment by iwpnd** [comment]
I created a first [draft](https://github.com/iwpnd/folium-vectortilelayer) for you that covers your basic use-case. Feel free to try out @giswqs. If that works fine for you, I can draft a PR with folium to incorporate into the plugin ecosystem.

**Comment by giswqs** [comment]
Awesome! Will try it out tonight. Thank you

**Comment by giswqs** [comment]
@iwpnd Nice! It works now. How do you determine `maxDetailZoom`? When it is great than 13, the layer won't show up. I guess different vector tiles will have a different `maxDetailZoom`. Is there a way to set an optimal `maxDetailZoom` programmitcally by default so that users don't have to specify it? Imagine that a user sets `maxDetailZoom` to 20, and the layer does not render. The user might not necessary keep trying a smaller `maxDetailZoom` until the layer shows up. 

![image](https://user-images.githubusercontent.com/5016453/217620403-6fb06554-da61-4992-9f86-679258eb443d.png)

**Comment by iwpnd** [comment]
A provider should offer the minimum and maximum bounds in their maps capabilities for the user to consider.
As for this leaflet plugin, the default `undefined` should be a good starting point and should render either way I suppose.

**Comment by giswqs** [comment]
If the `maxDetailZoom` is not specified, the layer does not render. So probably users need to figure it out by checking the metadata

```python
options = {
    'minDetailZoom': 5,
    # 'maxDetailZoom': 13,
    "style": { "weight": 3, "color": "green" }
}
```

**Comment by iwpnd** [comment]
Maybe this can be accounted for upstream in the plugin. The maintainer seems reactive on gitlab if you want to make your case. 
It most certainly makes sense to render to the point of what is offered by the provider, despite what is being set as `maxDetailZoom`.

**Comment by iwpnd** [comment]
There's also a `minZoom` / `maxZoom` property that is undocumented in Leaflet.VectorTileLayer. With the config below the layer shows up either way.

```python
ms_buildings = "https://planetarycomputer.microsoft.com/api/data/v1/vector/collections/ms-buildings/tilesets/global-footprints/tiles/{z}/{x}/{y}"

m = folium.Map(location=[40, -100], zoom_start=13)

options = {
    "minZoom": 5,
    "maxZoom": 18,
    'minDetailZoom': 5,
    'maxDetailZoom': 18,
    "style": { "weight": 3, "color": "green" }
}

vc = VectorTileLayer(ms_buildings,"world",options)

m.add_child(vc)
m
```

**Comment by iwpnd** [comment]
Did you have any chance to give this a try @giswqs?

**Comment by giswqs** [comment]
@iwpnd Yes, I tested it. I think the most critical parameter is `maxDetailZoom`. If it is larger than the max zoom level provided by source, the vector tile layer won't render. If it is smaller than the max zoom level provided by source, the polygons will be simplified. So users will have to do trial and error to figure out the optimal `maxDetailZoom`. 

The `maxZoom` parameter controls the zoom in button of the Zoom Control. It disables the zoom-in button when the map zoom level is larger than `maxZoom`. It does not really affect the vector tile rendering.

**Comment by iwpnd** [comment]
What do you suggest we continue here? I can create a release for you, or submit this to folium - or both.

**Comment by Conengmo** [comment]
Ideally we have only one plugin that provides vector tile support. If both libraries have a lot of overlap I'd rather not include both. Is the assumption both plugins do basically the same correct?

If true, I see two scenarios:
- Provide support for Leaflet.VectorTileLayer in a separate repo hosted by @iwpnd, which we'll link to in Folium.
- In Folium, swap out Leaflet.VectorGrid for Leaflet.VectorTileLayer.

Ideally our upstream plugins are robust, mature and well maintained.

@iwpnd I hope you can advice on whether it's good to swap out Leaflet.VectorGrid for Leaflet.VectorTileLayer, since you worked with both now.

**Comment by iwpnd** [comment]
> Is the assumption both plugins do basically the same correct?

This is correct. `Leaflet.VectorTileLayer` adds capability that `Leaflet.VectorGrid` is missing tho.

> If true, I see two scenarios

There is (almost) no downside to either approach, yet ultimately `Leaflet.VectorTileLayer` appears to have the better features and seems to be maintained (for now). The only downside I can think of is a breaking change in folium because of a plugin if you we're to swap one for the other.

**Comment by jkuebart** [comment]
> @iwpnd Yes, I tested it. I think the most critical parameter is `maxDetailZoom`. If it is larger than the max zoom level provided by source, the vector tile layer won't render. If it is smaller than the max zoom level provided by source, the polygons will be simplified. So users will have to do trial and error to figure out the optimal `maxDetailZoom`.
> 
> The `maxZoom` parameter controls the zoom in button of the Zoom Control. It disables the zoom-in button when the map zoom level is larger than `maxZoom`. It does not really affect the vector tile rendering.

I didn't follow this discussion too closely as I don't know about `folium`, but there shouldn't be much »trial and error« involved in setting the zoom parameters.

If the server serves tiles for the full range of `max/minZoom` levels, there's really no reason to use any of the other options. If the server lacks some zoom levels, `Leaflet.VectorTileLayer` offers two possible solutions:
 * Using `max/minNativeZoom`, the tiles on zoom levels outside the given range are loaded from the specified zoom level and scaled up or down including line widths and marker sizes. The result looks the same as it would with raster tiles without the obvious artifacts.
 * Using `max/minDetailZoom`, the tiles on zoom levels outside the given range are loaded from the specified zoom level and drawn larger or smaller while *keeping the same* line widths and marker sizes. The result looks as if the tile with the target zoom level was available, but obviously the amount of detail remains the same.

`max/minZoom` and `max/minNativeZoom` work as [documented for `L.GridLayer`](https://leafletjs.com/reference.html#gridlayer), to quote:
 * `maxZoom` The maximum zoom level up to which this layer will be displayed (inclusive).
 * `minZoom` The minimum zoom level down to which this layer will be displayed (inclusive).
 * `maxNativeZoom` Maximum zoom number the tile source has available. If it is specified, the tiles on all zoom levels higher than `maxNativeZoom` will be loaded from `maxNativeZoom` level and auto-scaled.
 * `minNativeZoom` Minimum zoom number the tile source has available. If it is specified, the tiles on all zoom levels lower than `minNativeZoom` will be loaded from `minNativeZoom` level and auto-scaled.

`Leaflet.VectorTileLayer` adds `max/minDetailZoom` which is [documented here](https://gitlab.com/jkuebart/Leaflet.VectorTileLayer/-/tree/main#leafletvectortilelayer).

**Comment by iwpnd** [comment]
https://pypi.org/project/folium-vectortilelayer/

for whoever is interested.

thank you @jkuebart for the plugin and your input here 🙏

**Comment by hansthen** [comment]
@iwpnd I was going through old issues and I found this. Thanks for your contribution. I will play around with it soon. I also updated the README to point to your plugin.
